client/state/data-layer/wpcom/jetpack-onboarding/index.js

```diff
@@ -116,13 +116,46 @@ export const saveJetpackOnboardingSettings = ( { dispatch, getState }, action )
 export const handleSaveSuccess = ( { dispatch }, { siteId, settings } ) =>
 	dispatch( saveJetpackOnboardingSettingsSuccess( siteId, settings ) );
 
-export const announceSaveFailure = ( { dispatch }, { siteId } ) =>
+export const announceSaveFailure = ( { dispatch }, { siteId } ) => {
 	dispatch(
 		errorNotice( translate( 'An unexpected error occurred. Please try again later.' ), {
 			id: `jpo-notice-error-${ siteId }`,
 			duration: 5000,
 		} )
 	);
+};
+
+export const retryOrAnnounceSaveFailure = ( { dispatch }, action ) => {
+	const MAX_WOOCOMMERCE_INSTALL_RETRIES = 2;
+	const { settings, siteId, type, meta: { dataLayer } } = action;
+	const { error, retryCount = 0 } = dataLayer;
+
+	// If we got a timeout on WooCommerce installation, try again (up to 3 times),
+	// since it might just be a slow server that actually ends up installing it
+	// properly, in which case a subsequent request will return 'success'.
+	if (
+		settings.installWooCommerce !== true ||
+		error.error !== 'http_request_failed' ||
```

Comment:
I don't think this is covering our use case well enough. This will inadvertently handle a bunch of other failure cases that fail with a `http_request_failed` error, for example (and that's not a full list):

* Could not open handle for fopen() to /path/to/file
* Too many redirects.
* Failed to write request to temporary file.
* A valid URL was not provided.
* User has blocked requests through HTTP.
* Destination directory for file streaming does not exist or is not writable.
* Any generic cURL error that came from `curl_error()`

I'd rather check for presence of `cURL error 28` in the error message, so we can only do this for cases when it actually times out.